dependencyResolutionManagement issue when plugins trigger repositories

### Expected Behavior

I would expect it to add defined repositories to all subprojects as the documentation says

### Current Behavior
It does not add any repositories to subprojects

### Context

See below

### Steps to Reproduce 
In our build we are adding this to the beginning of settings.gradle.kts

```kotlin
dependencyResolutionManagement {
  repositories {
    // Repos are declared roughly in order of likely to hit.

    // Google's maven central mirror: https://maven-central.storage.googleapis.com/index.html
    // To override this, set slack.gradle.config.customGoogleMavenCentralMirror in your user home gradle.properties to
    // something else.
    maven("https://maven-central.storage-download.googleapis.com/maven2/")
    google()
    gradlePluginPortal()
    mavenCentral()
    exclusiveContent {
      forRepository {
        maven {
          name = "JCenter"
          setUrl("https://jcenter.bintray.com/")
        }
      }
      filter {
        // Required for a dependency of Android lint.
        includeGroup("org.jetbrains.trove4j")
        // Required for Dokka
        includeModule("org.jetbrains.kotlinx", "kotlinx-html-jvm")
        includeGroup("org.jetbrains.dokka")
        includeModule("org.jetbrains", "markdown")
      }
    }
    // Kotlin EAPs, only tested on CI shadow jobs
    maven("https://dl.bintray.com/kotlin/kotlin-eap")
    @Suppress("UnstableApiUsage")
    exclusiveContent {
      forRepository {
        // For R8/D8 releases
        maven("https://storage.googleapis.com/r8-releases/raw")
      }
      filter {
        includeModule("com.android.tools", "r8")
      }
    }
  }
}
```

When trying to compile a simple kotlin gradle plugin subproject though, it fails instantly due to no repositories

```
Could not resolve compiler classpath. Check if Kotlin Gradle plugin repository is configured in project ':slack-plugin'.

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':slack-plugin:compileKotlin'.
> Could not resolve all files for configuration ':slack-plugin:kotlinCompilerClasspath'.
   > Could not find org.jetbrains.kotlin:kotlin-compiler-embeddable:1.4.20.
     Searched in the following locations:
       - file:/Users/zsweers/.m2/repository/org/jetbrains/kotlin/kotlin-compiler-embeddable/1.4.20/kotlin-compiler-embeddable-1.4.20.pom
     If the artifact you are trying to retrieve can be found in the repository but without metadata in 'Maven POM' format, you need to adjust the 'metadataSources { ... }' of the repository declaration.
     Required by:
         project :slack-plugin

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s

```

### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- A build scan https://scans.gradle.com/get-started is ideal -->
Build scan URL:

--- a/subprojects/core/src/main/java/org/gradle/api/internal/SettingsInternal.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/SettingsInternal.java
@@ -26,6 +26,7 @@ import org.gradle.caching.configuration.internal.BuildCacheConfigurationInternal
 import org.gradle.groovy.scripts.ScriptSource;
 import org.gradle.initialization.DefaultProjectDescriptor;
 import org.gradle.initialization.IncludedBuildSpec;
+import org.gradle.internal.management.DependencyResolutionManagementInternal;
 
 import java.io.File;
 import java.util.List;
@@ -70,4 +71,6 @@ public interface SettingsInternal extends Settings, PluginAwareInternal {
     BuildCacheConfigurationInternal getBuildCache();
 
     void preventFromFurtherMutation();
+
+    DependencyResolutionManagementInternal getDependencyResolutionManagement();
 }

--- a/subprojects/core/src/main/java/org/gradle/initialization/DefaultSettings.java
+++ b/subprojects/core/src/main/java/org/gradle/initialization/DefaultSettings.java
@@ -382,4 +382,9 @@ public abstract class DefaultSettings extends AbstractPluginAware implements Set
     public void preventFromFurtherMutation() {
         dependencyResolutionManagement.preventFromFurtherMutation();
     }
+
+    @Override
+    public DependencyResolutionManagementInternal getDependencyResolutionManagement() {
+        return dependencyResolutionManagement;
+    }
 }

--- a/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RepositoriesDeclaredInSettingsIntegrationTest.groovy
+++ b/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RepositoriesDeclaredInSettingsIntegrationTest.groovy
@@ -21,9 +21,12 @@ import org.gradle.integtests.fixtures.RequiredFeature
 import org.gradle.integtests.fixtures.ToBeFixedForConfigurationCache
 import org.gradle.test.fixtures.plugin.PluginBuilder
 import org.gradle.test.fixtures.server.http.MavenHttpPluginRepository
+import org.gradle.util.GradleVersion
 import org.gradle.util.TestPrecondition
 import org.gradle.util.ToBeImplemented
 import spock.lang.IgnoreIf
+import spock.lang.Issue
+
 // Restrict the number of combinations because that's not really what we want to test
 @RequiredFeature(feature = GradleMetadataResolveRunner.REPOSITORY_TYPE, value = "maven")
 @RequiredFeature(feature = GradleMetadataResolveRunner.GRADLE_METADATA, value = "true")
@@ -731,6 +734,31 @@ class RepositoriesDeclaredInSettingsIntegrationTest extends AbstractModuleDepend
 
         cleanup:
         pluginPortal.stop()
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/15336")
+    def "reasonable error message if a dependency cannot be resolved because local repositories differ"() {
+        buildFile << """
+            repositories {
+                maven {
+                    url "dummy"
+                }
+            }
+
+            dependencies {
+                conf 'org:module:1.0'
+            }
+
+        """
+
+        when:
+        fails ':checkDeps'
+
+        then:
+        failure.assertThatCause(containsNormalizedString("""Could not resolve all dependencies for configuration ':conf'.
+The project declares repositories, effectively ignoring the repositories you have declared in the settings.
+You can figure out how project repositories are declared by configuring your build to fail on project repositories.
+See https://docs.gradle.org/${GradleVersion.current().version}/userguide/declaring_repositories.html#sub:fail_build_on_project_repositories for details."""))
     }
 
     void withSettingsPlugin() {

--- a/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RepositoriesDeclaredInSettingsIntegrationTest.groovy
+++ b/subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RepositoriesDeclaredInSettingsIntegrationTest.groovy
@@ -737,6 +737,7 @@ class RepositoriesDeclaredInSettingsIntegrationTest extends AbstractModuleDepend
     }
 
     @Issue("https://github.com/gradle/gradle/issues/15336")
+    @ToBeFixedForConfigurationCache(because = "Decorated exception is not rethrown")
     def "reasonable error message if a dependency cannot be resolved because local repositories differ"() {
         buildFile << """
             repositories {

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationInternal.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationInternal.java
@@ -20,6 +20,7 @@ import org.gradle.api.artifacts.Configuration;
 import org.gradle.api.artifacts.DependencyConstraint;
 import org.gradle.api.artifacts.ExcludeRule;
 import org.gradle.api.artifacts.PublishArtifact;
+import org.gradle.api.artifacts.ResolveException;
 import org.gradle.api.internal.artifacts.ResolveContext;
 import org.gradle.api.internal.artifacts.transform.ExtraExecutionGraphDependenciesResolverFactory;
 import org.gradle.api.internal.attributes.AttributeContainerInternal;
@@ -94,6 +95,14 @@ public interface ConfigurationInternal extends ResolveContext, Configuration, De
     ConfigurationInternal getConsistentResolutionSource();
 
     Supplier<List<DependencyConstraint>> getConsistentResolutionConstraints();
+
+    /**
+     * Decorates a resolve exception with more context. This can be used
+     * to give hints to the user when a resolution error happens.
+     * @param e a resolve exception
+     * @return a decorated resolve exception, or the same exception
+     */
+    ResolveException maybeAddContext(ResolveException e);
 
     interface VariantVisitor {
         // The artifacts to use when this configuration is used as a configuration

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java
@@ -724,6 +724,26 @@ public class DefaultConfiguration extends AbstractFileCollection implements Conf
         return consistentResolutionConstraints;
     }
 
+    @Override
+    public ResolveException maybeAddContext(ResolveException e) {
+        if (ignoresSettingsRepositories()) {
+            return new ResolveExceptionWithHints(getDisplayName(), e.getCauses(),
+                "The project declares repositories, effectively ignoring the repositories you have declared in the settings.",
+                "You can figure out how project repositories are declared by configuring your build to fail on project repositories.",
+                "See " + documentationRegistry.getDocumentationFor("declaring_repositories", "sub:fail_build_on_project_repositories") + " for details.");
+        }
+        return e;
+    }
+
+    private boolean ignoresSettingsRepositories() {
+        if (owner instanceof ProjectInternal) {
+            ProjectInternal project = (ProjectInternal) this.owner;
+            return !project.getRepositories().isEmpty() &&
+                !project.getGradle().getSettings().getDependencyResolutionManagement().getRepositoryHandler().isEmpty();
+        }
+        return false;
+    }
+
     private void assertNoDependencyResolutionConsistencyCycle() {
         Set<ConfigurationInternal> sources = Sets.newLinkedHashSet();
         ConfigurationInternal src = this;

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java
@@ -686,7 +686,7 @@ public class DefaultConfiguration extends AbstractFileCollection implements Conf
                         projectPathString,
                         isVisible(),
                         isTransitive(),
-                        resolver.getRepositories()
+                        resolver::getRepositories
                     ));
             }
         });

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ResolveConfigurationResolutionBuildOperationDetails.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ResolveConfigurationResolutionBuildOperationDetails.java
@@ -21,6 +21,7 @@ import com.google.common.collect.ImmutableMap;
 import org.gradle.api.internal.artifacts.configurations.ResolveConfigurationDependenciesBuildOperationType.Repository;
 import org.gradle.api.internal.artifacts.repositories.ResolutionAwareRepository;
 import org.gradle.api.internal.artifacts.repositories.descriptor.RepositoryDescriptor;
+import org.gradle.internal.lazy.Lazy;
 import org.gradle.internal.operations.trace.CustomOperationTraceSerialization;
 import org.gradle.util.CollectionUtils;
 
@@ -31,6 +32,7 @@ import java.util.Collection;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.function.Supplier;
 
 class ResolveConfigurationResolutionBuildOperationDetails implements ResolveConfigurationDependenciesBuildOperationType.Details, CustomOperationTraceSerialization {
 
@@ -41,7 +43,7 @@ class ResolveConfigurationResolutionBuildOperationDetails implements ResolveConf
     private final String projectPath;
     private final boolean isConfigurationVisible;
     private final boolean isConfigurationTransitive;
-    private final List<Repository> repositories;
+    private final Supplier<List<Repository>> repositories;
 
     ResolveConfigurationResolutionBuildOperationDetails(
         String configurationName,
@@ -51,7 +53,7 @@ class ResolveConfigurationResolutionBuildOperationDetails implements ResolveConf
         @Nullable String projectPath,
         boolean isConfigurationVisible,
         boolean isConfigurationTransitive,
-        List<ResolutionAwareRepository> repositories
+        Supplier<List<ResolutionAwareRepository>> repositoriesSupplier
     ) {
         this.configurationName = configurationName;
         this.isScriptConfiguration = isScriptConfiguration;
@@ -60,7 +62,7 @@ class ResolveConfigurationResolutionBuildOperationDetails implements ResolveConf
         this.projectPath = projectPath;
         this.isConfigurationVisible = isConfigurationVisible;
         this.isConfigurationTransitive = isConfigurationTransitive;
-        this.repositories = RepositoryImpl.transform(repositories);
+        this.repositories = Lazy.unsafe().of(() -> RepositoryImpl.transform(repositoriesSupplier.get()));
     }
 
     @Override
@@ -101,7 +103,7 @@ class ResolveConfigurationResolutionBuildOperationDetails implements ResolveConf
 
     @Override
     public List<Repository> getRepositories() {
-        return repositories;
+        return repositories.get();
     }
 
     @Override
@@ -115,7 +117,7 @@ class ResolveConfigurationResolutionBuildOperationDetails implements ResolveConf
         model.put("configurationVisible", isConfigurationVisible);
         model.put("configurationTransitive", isConfigurationTransitive);
         ImmutableList.Builder<Object> repoBuilder = new ImmutableList.Builder<>();
-        for (Repository repository : repositories) {
+        for (Repository repository : repositories.get()) {
             ImmutableMap.Builder<String, Object> repoMapBuilder = new ImmutableMap.Builder<>();
             repoMapBuilder.put("id", repository.getId());
             repoMapBuilder.put("name", repository.getName());

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ResolveExceptionWithHints.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ResolveExceptionWithHints.java
@@ -1,0 +1,39 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.internal.artifacts.configurations;
+
+
+import org.gradle.api.artifacts.ResolveException;
+
+class ResolveExceptionWithHints extends ResolveException {
+
+    private final String[] hints;
+
+    public ResolveExceptionWithHints(String resolveContext, Iterable<? extends Throwable> causes, String... hints) {
+        super(resolveContext, causes);
+        this.hints = hints;
+    }
+
+    @Override
+    public String getMessage() {
+        StringBuilder sb = new StringBuilder(super.getMessage());
+        for (String hint : hints) {
+            sb.append("\n");
+            sb.append(hint);
+        }
+        return sb.toString();
+    }
+}

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ErrorHandlingConfigurationResolver.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ErrorHandlingConfigurationResolver.java
@@ -40,6 +40,7 @@ import org.gradle.api.internal.artifacts.repositories.ResolutionAwareRepository;
 import org.gradle.api.internal.attributes.AttributeContainerInternal;
 import org.gradle.api.internal.tasks.TaskDependencyResolveContext;
 import org.gradle.api.specs.Spec;
+import org.gradle.internal.resolve.ModuleVersionNotFoundException;
 
 import java.io.File;
 import java.util.List;
@@ -99,9 +100,20 @@ public class ErrorHandlingConfigurationResolver implements ConfigurationResolver
 
     private static ResolveException wrapException(Throwable e, ResolveContext resolveContext) {
         if (e instanceof ResolveException) {
-            return (ResolveException) e;
+            ResolveException resolveException = (ResolveException) e;
+            return maybeAddHintToResolveException(resolveContext, resolveException);
         }
         return new ResolveException(resolveContext.getDisplayName(), e);
+    }
+
+    private static ResolveException maybeAddHintToResolveException(ResolveContext resolveContext, ResolveException resolveException) {
+        List<? extends Throwable> causes = resolveException.getCauses();
+        boolean hasModuleNotFound = causes.stream().anyMatch(ModuleVersionNotFoundException.class::isInstance);
+        if (hasModuleNotFound && resolveContext instanceof ConfigurationInternal) {
+            ConfigurationInternal config = (ConfigurationInternal) resolveContext;
+            return config.maybeAddContext(resolveException);
+        }
+        return resolveException;
     }
 
     private static class ErrorHandlingLenientConfiguration implements LenientConfiguration {

--- a/subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationSpec.groovy
+++ b/subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationSpec.groovy
@@ -948,7 +948,6 @@ class DefaultConfigurationSpec extends Specification {
 
         then:
         interaction { resolveConfig(config) }
-        1 * resolver.getRepositories() >> []
         0 * resolver._
     }
 
@@ -1132,7 +1131,6 @@ class DefaultConfigurationSpec extends Specification {
         1 * resolver.resolveGraph(config, _) >> { ConfigurationInternal c, ResolverResults r ->
             r.graphResolved(Stub(ResolutionResult), Stub(ResolvedLocalComponentsResult), visitedArtifacts())
         }
-        1 * resolver.getRepositories() >> []
         0 * resolver._
     }
 
@@ -1173,7 +1171,6 @@ class DefaultConfigurationSpec extends Specification {
         1 * resolver.resolveGraph(config, _) >> { ConfigurationInternal c, ResolverResults r ->
             r.graphResolved(Stub(ResolutionResult), Stub(ResolvedLocalComponentsResult), visitedArtifacts())
         }
-        1 * resolver.getRepositories() >> []
         0 * resolver._
 
         when:
@@ -1223,7 +1220,6 @@ class DefaultConfigurationSpec extends Specification {
         1 * resolver.resolveArtifacts(config, _) >> { ConfigurationInternal c, ResolverResults r ->
             r.artifactsResolved(Stub(ResolvedConfiguration), visitedArtifacts())
         }
-        1 * resolver.getRepositories() >> []
         0 * resolver._
     }
 
@@ -1247,7 +1243,6 @@ class DefaultConfigurationSpec extends Specification {
         1 * resolver.resolveArtifacts(config, _) >> { ConfigurationInternal c, ResolverResults r ->
             r.artifactsResolved(Stub(ResolvedConfiguration), visitedArtifacts())
         }
-        1 * resolver.getRepositories() >> []
         0 * resolver._
 
         when:

--- a/subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_repositories.adoc
+++ b/subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_repositories.adoc
@@ -504,6 +504,8 @@ include::sample[dir="snippets/artifacts/defineRepositoryInSettings/groovy",files
 include::sample[dir="snippets/artifacts/defineRepositoryInSettings/kotlin",files="settings.gradle.kts[tags=prefer_settings]"]
 ====
 
+[[sub:fail_build_on_project_repositories]]
+
 If, for some reason, a project or a plugin declares a repository in a project, Gradle would warn you.
 You can however make it _fail the build_ if you want to enforce that only settings repositories are used:
 

